pkg/volume/snapshot_params.go

```diff
@@ -0,0 +1,98 @@
+package volume
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/pborman/uuid"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/piraeusdatastore/linstor-csi/pkg/linstor"
+)
+
+//go:generate enumer -type=SnapshotType -trimprefix=SnapshotType
+
+type SnapshotType int
+
+const (
+	SnapshotTypeInCluster SnapshotType = iota
+	SnapshotTypeS3
+	SnapshotTypeLinstor
+)
+
+type SnapshotParameter struct {
+	Type             SnapshotType `json:"type,omitempty"`
+	AllowIncremental bool         `json:"allow-incremental"`
+	RemoteName       string       `json:"remote-name,omitempty"`
+	S3Endpoint       string       `json:"s3-endpoint,omitempty"`
+	S3Bucket         string       `json:"s3-bucket,omitempty"`
+	S3SigningRegion  string       `json:"s3-signing-region,omitempty"`
+	S3UsePathStyle   bool         `json:"s3-use-path-style"`
+	S3AccessKey      string       `json:"-"`
+	S3SecretKey      string       `json:"-"`
+}
+
+func NewSnapshotParameter(params, secrets map[string]string) (*SnapshotParameter, error) {
+	p := &SnapshotParameter{}
+
+	for k, v := range params {
+		if !strings.HasPrefix(k, linstor.SnapshotParameterNamespace) {
+			log.WithField("key", k).Debug("skipping parameter without snapshot parameter prefix")
+			continue
+		}
+
+		k = k[len(linstor.SnapshotParameterNamespace):]
+
+		switch k {
+		case "/type":
+			t, err := SnapshotTypeString(v)
+			if err != nil {
+				return nil, err
+			}
+
+			p.Type = t
+		case "/allow-incremental":
+			b, err := strconv.ParseBool(v)
+			if err != nil {
+				return nil, err
+			}
+
+			p.AllowIncremental = b
+		case "/remote-name":
+			p.RemoteName = v
+		case "/s3-endpoint":
+			p.S3Endpoint = v
+		case "/s3-bucket":
+			p.S3Bucket = v
+		case "/s3-signing-region":
+			p.S3SigningRegion = v
+		case "/s3-use-path-style":
+			b, err := strconv.ParseBool(v)
+			if err != nil {
+				return nil, err
+			}
+
+			p.S3UsePathStyle = b
+		default:
+			log.WithField("key", k).Warn("ignoring unknown snapshot parameter key")
+		}
+	}
+
+	if p.Type != SnapshotTypeInCluster && p.RemoteName == "" {
+		// encode params as json, so we can have a stable hash
```

Comment:
Do we actually want to hash the entire `SnapshotParameter`? Only certain parameters really define the remote. Although, perhaps only `AllowIncremental` is really specific to the snapshot.

Reply:
Maybe we should just require a name to be specified...

Reply:
Replaced by requiring a remote name for S3 snapshots.